entity-framework/core/internals/internal-service-provider.md

```diff
@@ -0,0 +1,166 @@
+
+### Registration of services
+
+[`EntityFrameworkServicesBuilder.TryAddCoreServices`] is used by database providers to register default service implementations.
+As the name suggests, TryAddCoreServices only register a default implementation if some other implementation has not already been registered.
+This means database providers can override service implementations by registering an implementation before calling TryAddCoreServices.
+
+Database providers register overrides using [`EntityFrameworkServicesBuilder.TryAdd`]() methods.
+These methods ensure that provider registrations use the service characteristics (notably, lifetime) defined by EF Core.
+
+For example, [`AddEntityFrameworkCosmos`]() looks something like this:
+
+```CSharp
+public static IServiceCollection AddEntityFrameworkCosmos([NotNull] this IServiceCollection serviceCollection)
+{
+    var builder = new EntityFrameworkServicesBuilder(serviceCollection)
+        .TryAdd<LoggingDefinitions, CosmosLoggingDefinitions>()
+        .TryAdd<IDatabaseProvider, DatabaseProvider<CosmosOptionsExtension>>()
+        // ... more Cosmos service registrations.
+    
+    builder.TryAddCoreServices();
+    
+    return serviceCollection;
+}
+```
+
+### Relational database providers
+
+Database providers for relational databases share many common services.
+These services are defined by [`EntityFrameworkRelationalServicesBuilder`](), which inherits from EntityFrameworkServicesBuilder.
+EntityFrameworkRelationalServicesBuilder overrides TryAddCoreServices to register relational implementations before calling the base method to register EF Core defaults.
+
+A relational database provider uses the same code shown above, except using a EntityFrameworkRelationalServicesBuilder instance.
+For example, see [`AddEntityFrameworkSqlServer`()].
+This means that:
+* Provider-specific services are registered first.
+* Relational-specific services are then registered unless the provider has already registered a provider-specific implementation.
+* Default services are then registered only when a relational-specific or provider-specific implementation has not been already registered.
+
+### Extending services
+
+Sometimes it is useful for a provider or relational-specific service to extend the core service interface.
+For example, [`IRelationalTypeMappingSource`]() inherits from [`ITypeMappingSource`]() adding relational-specific mapping methods.
+
+Non-relational code will resolve this service using ITypeMappingSource.
+Relational code that needs the relational-specific methods will resolve the same service using IRelationalTypeMappingSource.
+Both these resolutions must return _the same service instance_.
+This is achieved redirecting resolution for ITypeMappingSource to use IRelationalTypeMappingSource:
+
+```CSharp
+TryAdd<ITypeMappingSource>(p => p.GetService<IRelationalTypeMappingSource>());
+```
+
+## Resolving services
+
+Service implementations can be public or [public but internal]().
```

Comment:
Maybe introduce the idea of pubternal before, including some of the reasoning (in a more general intro not specific to the internal service provider?)

Reply:
The the dummy link here will ultimately go to a page on that.